broker.go

```diff
@@ -0,0 +1,609 @@
+// of messages.  Handlers do not have persistent goroutines, but start a new
+// goroutine to handle a single message.
+type Broker struct {
+	// dss captures the default selector sequence passed to
+	// ExploreRecursiveWithStopNode.
+	dss  ipld.Node
+	host host.Host
+	lsys ipld.LinkSystem
+
+	addrTTL   time.Duration
+	psub      *pubsub.Subscription
+	topic     *pubsub.Topic
+	topicName string
+
+	allowPeer     AllowPeerFunc
+	handlers      map[peer.ID]*handler
+	handlersMutex sync.Mutex
+
+	// inEvents is used to send a SyncFinished from a peer handler to the
+	// distributeEvents goroutine.
+	inEvents chan SyncFinished
+
+	// outEventsChans is a slice of channels, where each channel delivers a
+	// copy of a SyncFinished to an OnSyncFinished reader.
+	outEventsChans []chan SyncFinished
+	outEventsMutex sync.Mutex
+
+	// closing signals that the Broker is closing.
+	closing chan struct{}
+	// cancelps cancels pubsub.
+	cancelps context.CancelFunc
+	// closeOnde ensures that the Close only happens once.
+	closeOnce sync.Once
+
+	dtSync   *dtsync.Sync
+	httpSync *httpsync.Sync
+}
+
+// SyncFinished notifies an OnSyncFinished reader that a specified peer
+// completed a sync.  The channel receives events from providers that are
+// manually synced to the latest, as well as those auto-discovered.
+type SyncFinished struct {
+	// Cid is the CID identifying the link that finished and is now the latest
+	// sync for a specific peer.
+	Cid cid.Cid
+	// PeerID identifies the peer this SyncFinished event pertains to.
+	PeerID peer.ID
+}
+
+// handler holds state that is specific to a peer
+type handler struct {
+	broker *Broker
+	// distEvents is used to communicate a syncDone event back to the Broker
+	// for distribution to OnSyncFinished readers.
+	distEvents chan<- SyncFinished
+	latestSync ipld.Link
+	msgChan    chan cid.Cid
+	// peerID is the ID of the peer this handler is responsible for.
+	peerID peer.ID
+	// syncMutex serializes the handling of individual syncs
+	syncMutex sync.Mutex
+}
+
+// NewBroker creates a new Broker that process pubsub messages.
+func NewBroker(host host.Host, ds datastore.Batching, lsys ipld.LinkSystem, topic string, dss ipld.Node, options ...Option) (*Broker, error) {
+	cfg := config{
+		addrTTL: defaultAddrTTL,
+	}
+	err := cfg.apply(options)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, cancelPubsub := context.WithCancel(context.Background())
+
+	var pubsubTopic *pubsub.Topic
+	if cfg.topic == nil {
+		pubsubTopic, err = gpubsub.MakePubsub(ctx, host, topic)
+		if err != nil {
+			cancelPubsub()
+			return nil, err
+		}
+		cfg.topic = pubsubTopic
+	}
+	psub, err := cfg.topic.Subscribe()
+	if err != nil {
+		cancelPubsub()
+		return nil, err
+	}
+
+	dtSync, err := dtsync.NewSync(host, ds, lsys, cfg.dtManager)
+	if err != nil {
+		cancelPubsub()
+		return nil, err
+	}
+
+	lb := &Broker{
+		dss:  dss,
+		host: host,
+		lsys: lsys,
+
+		addrTTL:   cfg.addrTTL,
+		psub:      psub,
+		topic:     pubsubTopic,
```

Comment:
Should this be `cfg.topic` instead? Otherwise this will be nil if `cfg.topic` is set (via option)

Reply:
yes